src/dispdrivers/vgacontroller.cpp

```diff
@@ -151,33 +151,95 @@ void IRAM_ATTR VGAController::VSyncInterrupt(void * arg)
+{
+  switch (mode) {
+    case PaintMode::Set:
+      return [&] (int Y, int X1, int X2, uint8_t pattern) { rawFillRow(Y, X1, X2, pattern); };
+    case PaintMode::OR:
+    case PaintMode::ORNOT:
+      return [&] (int Y, int X1, int X2, uint8_t pattern) { rawORRow(Y, X1, X2, pattern); };
+    case PaintMode::AND:
+    case PaintMode::ANDNOT:
+      return [&] (int Y, int X1, int X2, uint8_t pattern) { rawANDRow(Y, X1, X2, pattern); };
+    case PaintMode::XOR:
+      return [&] (int Y, int X1, int X2, uint8_t pattern) { rawXORRow(Y, X1, X2, pattern); };
+    case PaintMode::Invert:
+      return [&] (int Y, int X1, int X2, uint8_t pattern) { rawInvertRow(Y, X1, X2); };
+    default:  // PaintMode::NoOp
+      return [&] (int Y, int X1, int X2, uint8_t pattern) { return; };
+  }
+}
+
+
+
 void IRAM_ATTR VGAController::setPixelAt(PixelDesc const & pixelDesc, Rect & updateRect)
 {
-  genericSetPixelAt(pixelDesc, updateRect,
-                    [&] (RGB888 const & color)          { return preparePixel(color); },
-                    [&] (int X, int Y, uint8_t pattern) { VGA_PIXEL(X, Y) = pattern; }
-                   );
+  auto paintMode = paintState().paintOptions.mode;
+  genericSetPixelAt(pixelDesc, updateRect, getPixelLamda(paintMode), setPixelLamda(paintMode));
 }
 
 
+
 // coordinates are absolute values (not relative to origin)
 // line clipped on current absolute clipping rectangle
 void IRAM_ATTR VGAController::absDrawLine(int X1, int Y1, int X2, int Y2, RGB888 color)
 {
+  auto paintMode = paintState().paintOptions.mode;
+
   genericAbsDrawLine(X1, Y1, X2, Y2, color,
-                     [&] (RGB888 const & color)                   { return preparePixel(color); },
-                     [&] (int Y, int X1, int X2, uint8_t pattern) { rawFillRow(Y, X1, X2, pattern); },
-                     [&] (int Y, int X1, int X2)                  { rawInvertRow(Y, X1, X2); },
-                     [&] (int X, int Y, uint8_t pattern)          { VGA_PIXEL(X, Y) = pattern; },
-                     [&] (int X, int Y)                           { VGA_INVERT_PIXEL(X, Y); }
+                     getPixelLamda(paintMode),
+                     fillRowLamda(paintMode),
+                     [&] (int Y, int X1, int X2) { rawInvertRow(Y, X1, X2); },
+                     setPixelLamda(paintMode),
+                     [&] (int X, int Y)          { VGA_INVERT_PIXEL(X, Y); }
                      );
 }
 
 
 // parameters not checked
 void IRAM_ATTR VGAController::rawFillRow(int y, int x1, int x2, RGB888 color)
 {
-  rawFillRow(y, x1, x2, preparePixel(color));
+  // This version, passing an RGB888 color, is only used by shape drawing methods,
+  // so we will pick fill method based on paint mode
+  auto paintMode = paintState().paintOptions.mode;
+  auto getPixel = getPixelLamda(paintMode);
+  auto pixel = getPixel(color);
+  auto fill = fillRowLamda(paintMode);
+  fill(y, x1, x2, pixel);
```

Comment:
changing this particular function to use the lambdas ensures that essentially all "fill" drawing operations will now obey the paint mode that has been set

this relies on the fact that the current codebase uses `rawFillRow` passing in an RGB888 colour _exclusively_ for those operations.

ideally, the name of this function should change to remove the ambiguity - that becomes practical once the other vga controller variants support the new plot modes

Reply:
this version of `rawFillRow` got renamed to be `fillRow`